File: benchmarks/tsb/bench_multi_index_sort_equals.ts
/**
 * Benchmark: MultiIndex.sortValues() and equals() on 100k-pair MultiIndex
 */
import { MultiIndex } from "../../src/index.ts";

const ROWS = 100_000;
const WARMUP = 3;
const ITERATIONS = 10;

const a = Array.from({ length: ROWS }, (_, i) => `a${i % 100}`);
const b = Array.from({ length: ROWS }, (_, i) => i % 1000);
const tuples: [string, number][] = a.map((v, i) => [v, b[i]]);
const mi = new MultiIndex({ tuples });
const mi2 = new MultiIndex({ tuples: tuples.slice() });

for (let i = 0; i < WARMUP; i++) {
  mi.sortValues();
  mi.equals(mi2);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  mi.sortValues();
  mi.equals(mi2);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "multi_index_sort_equals",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_multi_index_sort_equals.py
"""Benchmark: MultiIndex sort_values and equals on 100k-pair MultiIndex"""
import json, time
import pandas as pd

ROWS = 100_000
WARMUP = 3
ITERATIONS = 10

a = [f"a{i % 100}" for i in range(ROWS)]
b = [i % 1000 for i in range(ROWS)]
tuples = list(zip(a, b))

mi = pd.MultiIndex.from_tuples(tuples)
mi2 = pd.MultiIndex.from_tuples(tuples[:])

for _ in range(WARMUP):
    mi.sort_values()
    mi.equals(mi2)

start = time.perf_counter()
for _ in range(ITERATIONS):
    mi.sort_values()
    mi.equals(mi2)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "multi_index_sort_equals",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
